File: ios/FamilyApp/Features/Birthday/BirthdayScreen.swift
// Birthdays — the iOS twin of BirthdayScreen.kt: countdown cards sorted by next
// occurrence, "Turning N" chip, urgency-tinted days pill, add/edit sheets.
import SwiftUI

struct BirthdayScreen: View {
    let viewModel: BirthdayViewModel

    @State private var showAdd = false
    @State private var editing: BirthdayModel?

    var body: some View {
        let today = LocalDate.today()
        let sorted = sortedByNextBirthday(viewModel.birthdays, today: today)

        ZStack(alignment: .bottomTrailing) {
            Group {
                if viewModel.isLoading {
                    LoadingState().frame(maxHeight: .infinity, alignment: .top)
                } else if sorted.isEmpty {
                    EmptyState(
                        systemImage: "birthday.cake.fill",
                        title: L("No birthdays"),
                        subtitle: L("Add family birthdays so you never miss a celebration.")
                    )
                } else {
                    List {
                        ForEach(sorted) { birthday in
                            BirthdayCard(birthday: birthday, today: today) {
                                // Only the creator may edit (auto-birthdays belong to the person).
                                if birthday.madeByUserId == viewModel.currentUserId { editing = birthday }
                            }
                                .listRowSeparator(.hidden)
                                .listRowBackground(Color.clear)
                                .listRowInsets(EdgeInsets(
                                    top: 6, leading: Spacing.screenEdge,
                                    bottom: 6, trailing: Spacing.screenEdge
                                ))
                                .swipeActions(edge: .trailing) {
                                    Button(role: .destructive) {
                                        viewModel.delete(birthday)
                                    } label: {
                                        Label("Delete", systemImage: "trash")
                                    }
                                }
                        }
                    }
                    .listStyle(.plain)
                    .scrollContentBackground(.hidden)
                    .refreshable { viewModel.refresh() }
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)

            FloatingActionButton(text: L("Add birthday"), systemImage: "plus") { showAdd = true }
        }
        .ambientBackground()
        .featureTopBar(L("Birthdays"))
        .resumeEffect { viewModel.refresh() }
        .sheet(isPresented: $showAdd) {
            BirthdaySheet(title: L("Add birthday"), confirmLabel: L("Add")) { name, date, icon, color in
                viewModel.add(name: name, date: date, icon: icon, color: color)
                showAdd = false
            }
        }
        .sheet(item: $editing) { birthday in
            BirthdaySheet(
                title: L("Edit birthday"),
                confirmLabel: L("Save"),
                initialName: birthday.name,
                initialDate: birthday.date,
                initialIcon: birthday.icon,
                initialColor: birthday.color
            ) { name, date, icon, color in
                viewModel.update(id: birthday.id, name: name, date: date, icon: icon, color: color)
                editing = nil
            }
        }
    }
}

private struct BirthdayCard: View {
    let birthday: BirthdayModel
    let today: LocalDate
    let onEdit: () -> Void

    var body: some View {
        let next = nextBirthdayDate(birthday.date, today: today)
        let age = turnsAge(birthday.date, today: today)
        let daysUntil = next.map { today.daysUntil($0) }
        // Show the actual date of birth (year included), not the next occurrence.
        let displayDate = LocalDate(iso: birthday.date).map(formatFullDate) ?? birthday.date
        let isToday = daysUntil == 0
        let accent = birthdayAccent(birthday)

        Button(action: onEdit) {
            HStack(spacing: 12) {
                Circle()
                    .fill(accent.opacity(0.16))
                    .frame(width: 46, height: 46)
                    .overlay(
                        Image(systemName: IconKeyMap.calendarSymbol(birthday.icon))
                            .font(.system(size: 20, weight: .semibold))
                            .foregroundStyle(accent)
                    )
                VStack(alignment: .leading, spacing: 3) {
                    Text(birthday.name)
                        .font(.system(size: 15.5, weight: .semibold))
                        .foregroundStyle(Color.appOnSurface)
                    HStack(spacing: 5) {
                        Text(displayDate)
                            .foregroundStyle(Color.appCaption)
                        if let age {
                            Text("· turning \(age)")
                                .foregroundStyle(FeatureAccent.birthdays.stroke)
                        }
                    }
                    .font(.system(size: 12.5, weight: .medium))
                }
                Spacer()
                if let daysUntil {
                    daysPill(daysUntil)
                }
            }
            .padding(Spacing.cardPadding)
            .modifier(BirthdaySurface(isToday: isToday))
        }
        .buttonStyle(PressScaleButtonStyle())
        .accessibilityLabel("\(birthday.name)'s birthday, \(displayDate)")
    }

    @ViewBuilder
    private func daysPill(_ days: Int) -> some View {
        switch days {
        case 0:
            Text("Today! 🎉")
                .font(.system(size: 12, weight: .bold))
                .foregroundStyle(.white)
                .padding(.horizontal, 11).padding(.vertical, 5)
                .background(Color.appSuccess, in: Capsule())
                .shadow(color: Color.appSuccess.opacity(0.35), radius: 8, y: 3)
        case 1...7:
            Text("In \(days) days")
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(Palette.weekAmberText)
                .padding(.horizontal, 10).padding(.vertical, 5)
                .background(Palette.amber500.opacity(0.16), in: Capsule())
        default:
            Text("In \(days) days")
                .font(.system(size: 12, weight: .medium))
                .foregroundStyle(Color.appCaption)
                .padding(.horizontal, 10).padding(.vertical, 5)
                .background(Color(hex: 0x8B92AC).opacity(0.12), in: Capsule())
        }
    }

    /// Today = glass + green ring; all other (future/this week) = solid glass card.
    private struct BirthdaySurface: ViewModifier {
        let isToday: Bool
        func body(content: Content) -> some View {
            if isToday {
                content
                    .glassCard(cornerRadius: Radius.overviewCard)
                    .overlay(
                        RoundedRectangle(cornerRadius: Radius.overviewCard, style: .continuous)
                            .strokeBorder(Color.appSuccess.opacity(0.5), lineWidth: 1.5)
                    )
                    .shadow(color: Color.appSuccess.opacity(0.18), radius: 12, y: 8)
            } else {
                content.rowSurface(ghost: false, cornerRadius: Radius.overviewCard)
            }
        }
    }

    private func formatFullDate(_ date: LocalDate) -> String {
        let instant = Date(timeIntervalSince1970: TimeInterval(date.epochDay) * 86400)
        let formatter = DateFormatter()
        formatter.dateStyle = .long
        formatter.locale = appLocale
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter.string(from: instant)
    }
}

/// A birthday's badge colour — the user-picked colour, else the birthdays accent.
func birthdayAccent(_ birthday: BirthdayModel) -> Color {
    if let hex = birthday.color { return Color(hex: UInt32(truncatingIfNeeded: hex)) }
    return FeatureAccent.birthdays.stroke
}

private struct BirthdaySheet: View {
    let title: String
    let confirmLabel: String
    var initialName = ""
    var initialDate = ""
    var initialIcon = "cake"
    var initialColor: Int?
    let onConfirm: (String, String, String, Int?) -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var name = ""
    @State private var date = ""
    @State private var icon = "cake"
    @State private var color: Int?

    private var canConfirm: Bool {
        !name.trimmingCharacters(in: .whitespaces).isEmpty && !date.isEmpty
    }

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.lg) {
            SheetHeader(title: title, confirmTitle: confirmLabel, confirmEnabled: canConfirm) {
                dismiss()
            } onConfirm: {
                onConfirm(name.trimmingCharacters(in: .whitespaces), date, icon, color)
                dismiss()
            }
            GlassField(systemImage: "person", placeholder: L("Name"), text: $name)
            BirthdayPickerField(isoDate: $date, label: L("Birthday *"))

            SectionHeader(text: L("Icon"))
            IconGrid(
                options: IconOptions.calendar,
                selected: icon,
                symbolFor: IconKeyMap.calendarSymbol
            ) { icon = $0 }

            SectionHeader(text: L("Color"))
            EventColorPicker(selection: $color)
        }
        .padding(.horizontal, Spacing.screenEdge)
        .padding(.top, Spacing.lg)
        .padding(.bottom, Spacing.xl)
        .huggingSheet()
        .onAppear {
            name = initialName
            date = initialDate
            icon = initialIcon
            color = initialColor
        }
    }
}

File: android/app/src/main/java/com/sandnes/familyapp/ui/birthday/BirthdayScreen.kt
@file:Suppress("ktlint:standard:function-naming")

package com.sandnes.familyapp.ui.birthday

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Cake
import androidx.compose.material.icons.filled.Person
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.sandnes.familyapp.R
import com.sandnes.familyapp.data.BirthdayModel
import com.sandnes.familyapp.ui.components.AppFab
import com.sandnes.familyapp.ui.components.BirthdayPickerField
import com.sandnes.familyapp.ui.components.ColorPickerRow
import com.sandnes.familyapp.ui.components.CreationSheet
import com.sandnes.familyapp.ui.components.EmptyState
import com.sandnes.familyapp.ui.components.FeatureTopBar
import com.sandnes.familyapp.ui.components.IconGrid
import com.sandnes.familyapp.ui.components.ListSkeleton
import com.sandnes.familyapp.ui.components.PillTag
import com.sandnes.familyapp.ui.components.PullRefresh
import com.sandnes.familyapp.ui.components.RefreshOnResume
import com.sandnes.familyapp.ui.components.SheetField
import com.sandnes.familyapp.ui.components.SwipeToRevealDelete
import com.sandnes.familyapp.ui.theme.AppColorPalette
import com.sandnes.familyapp.ui.theme.FeatureAccent
import com.sandnes.familyapp.ui.theme.IconKeyMap
import com.sandnes.familyapp.ui.theme.IconOptions
import com.sandnes.familyapp.ui.theme.Radius
import com.sandnes.familyapp.ui.theme.Spacing
import com.sandnes.familyapp.ui.theme.glassCard
import com.sandnes.familyapp.ui.theme.hexColor
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

private val BIRTH_DATE_FMT =
    DateTimeFormatter.ofLocalizedDate(java.time.format.FormatStyle.LONG).withLocale(Locale.getDefault())

@Composable
fun BirthdayScreen(
    onBack: () -> Unit,
    viewModel: BirthdayViewModel = hiltViewModel(),
) {
    val birthdays by viewModel.birthdays.collectAsStateWithLifecycle(emptyList())
    val isLoading by viewModel.isLoading.collectAsStateWithLifecycle(false)
    val currentUserId by viewModel.currentUserId.collectAsStateWithLifecycle(null)
    var showAdd by remember { mutableStateOf(false) }
    var editing by remember { mutableStateOf<BirthdayModel?>(null) }

    RefreshOnResume { viewModel.refresh() }

    val today = remember { LocalDate.now() }
    // List stays sorted by the next upcoming occurrence, even though each card shows the birth date.
    val sorted =
        remember(birthdays) {
            birthdays.sortedWith { a, b ->
                val nextA = nextBirthdayDate(a.date, today) ?: LocalDate.MAX
                val nextB = nextBirthdayDate(b.date, today) ?: LocalDate.MAX
                nextA.compareTo(nextB)
            }
        }

    Scaffold(
        containerColor = Color.Transparent,
        topBar = { FeatureTopBar(stringResource(R.string.birthdays), onBack) },
        floatingActionButton = {
            AppFab(text = stringResource(R.string.add_birthday), icon = Icons.Filled.Add, onClick = { showAdd = true })
        },
    ) { padding ->
        PullRefresh(
            onRefresh = { viewModel.refresh().join() },
            modifier = Modifier.fillMaxSize().padding(padding),
        ) {
            if (isLoading) {
                ListSkeleton(Modifier.fillMaxSize())
            } else if (sorted.isEmpty()) {
                Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                    EmptyState(
                        Icons.Filled.Cake,
                        stringResource(R.string.no_birthdays),
                        stringResource(R.string.add_family_birthdays_so_you_never_miss_a_celebration),
                        actionLabel = stringResource(R.string.add_birthday),
                        onAction = { showAdd = true },
                    )
                }
            } else {
                LazyColumn(
                    Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(Spacing.screenEdge),
                    verticalArrangement = Arrangement.spacedBy(12.dp),
                ) {
                    items(sorted, key = { it.id }) { b ->
                        // Only the creator may edit or delete (auto-birthdays belong to the person; RLS enforces).
                        val canEdit = b.madeByUserId == currentUserId
                        if (canEdit) {
                            SwipeToRevealDelete(
                                onDelete = { viewModel.delete(b) },
                                modifier = Modifier.animateItem(),
                                shape = RoundedCornerShape(Radius.overviewCard),
                            ) {
                                BirthdayCard(b, today, onEdit = { editing = b })
                            }
                        } else {
                            BirthdayCard(b, today, onEdit = null, modifier = Modifier.animateItem())
                        }
                    }
                    item { Spacer(Modifier.height(80.dp)) }
                }
            }
        }
    }

    if (showAdd) {
        BirthdayDialog(
            title = stringResource(R.string.add_birthday),
            confirmLabel = stringResource(R.string.add),
            onDismiss = { showAdd = false },
            onConfirm = { name, date, icon, color ->
                viewModel.add(name, date, icon, color)
                showAdd = false
            },
        )
    }

    editing?.let { birthday ->
        BirthdayDialog(
            title = stringResource(R.string.edit_birthday),
            confirmLabel = stringResource(R.string.save),
            initialName = birthday.name,
            initialDate = birthday.date,
            initialIcon = birthday.icon,
            initialColor = birthday.color,
            onDismiss = { editing = null },
            onConfirm = { name, date, icon, color ->
                viewModel.update(birthday.id, name, date, icon, color)
                editing = null
            },
        )
    }
}

@Composable
private fun BirthdayCard(
    b: BirthdayModel,
    today: LocalDate,
    onEdit: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    val nextDate = nextBirthdayDate(b.date, today)
    val age = turnsAge(b.date, today)
    val daysUntil = nextDate?.let { (it.toEpochDay() - today.toEpochDay()).toInt() }
    // Show the actual date of birth (year included), not the next occurrence.
    val displayDate =
        runCatching { LocalDate.parse(b.date).format(BIRTH_DATE_FMT) }.getOrDefault(b.date)
    val isToday = daysUntil == 0
    val accent = hexColor(b.color) ?: FeatureAccent.Birthdays.stroke()

    val daysLabel =
        when {
            daysUntil == 0 -> stringResource(R.string.today_exclaim)
            daysUntil != null -> stringResource(R.string.in_days, daysUntil)
            else -> ""
        }
    val cardDescription =
        "${b.name}'s birthday, $displayDate${if (daysLabel.isNotEmpty()) ", $daysLabel" else ""}"

    // Today's birthday gets a celebratory green ring; every card is a glass surface.
    val base =
        modifier
            .fillMaxWidth()
            .glassCard(Radius.overviewCard)
            .then(
                if (isToday) {
                    Modifier.border(1.5.dp, Color(0xFF22C55E).copy(alpha = 0.5f), RoundedCornerShape(Radius.overviewCard))
                } else {
                    Modifier
                },
            ).then(if (onEdit != null) Modifier.clickable { onEdit() } else Modifier)
            .padding(Spacing.cardPadding)
            .semantics { contentDescription = cardDescription }

    Row(base, verticalAlignment = Alignment.CenterVertically) {
        // Circular badge — accent at 16% fill with the accent glyph (mirrors the iOS row).
        Box(
            Modifier
                .size(46.dp)
                .clip(CircleShape)
                .background(accent.copy(alpha = 0.16f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                IconKeyMap.birthday(b.icon),
                contentDescription = null,
                tint = accent,
                modifier = Modifier.size(22.dp),
            )
        }
        Spacer(Modifier.size(14.dp))
        Column(Modifier.weight(1f)) {
            Text(
                b.name,
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
                fontWeight = FontWeight.SemiBold,
            )
            // FlowRow so "turning N" wraps as a whole chunk instead of mid-phrase when the
            // localized date + countdown pill leave little width.
            androidx.compose.foundation.layout.FlowRow {
                Text(
                    displayDate,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                if (age != null) {
                    Text(
                        " " + stringResource(R.string.turning_age, age),
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Medium,
                        color = accent,
                        maxLines = 1,
                    )
                }
            }
        }
        if (daysUntil != null) {
            Spacer(Modifier.size(8.dp))
            DaysPill(daysUntil)
        }
    }
}

/** Urgency-tinted countdown pill: Today (green), within a week (amber), otherwise muted. */
@Composable
private fun DaysPill(days: Int) {
    when (birthdayUrgency(days)) {
        BirthdayUrgency.TODAY ->
            PillTag(
                text = stringResource(R.string.today_exclaim) + " 🎉",
                container = Color(0xFF22C55E),
                content = Color.White,
            )
        BirthdayUrgency.SOON ->
            PillTag(
                text = stringResource(R.string.in_days, days),
                container = Color(0xFFF59E0B).copy(alpha = 0.18f),
                content = Color(0xFFB45309),
            )
        BirthdayUrgency.LATER ->
            PillTag(
                text = stringResource(R.string.in_days, days),
                container = MaterialTheme.colorScheme.surfaceVariant,
                content = MaterialTheme.colorScheme.onSurfaceVariant,
            )
    }
}

@Composable
private fun BirthdayDialog(
    title: String,
    confirmLabel: String,
    initialName: String = "",
    initialDate: String = "",
    initialIcon: String = "cake",
    initialColor: Int? = null,
    onDismiss: () -> Unit,
    onConfirm: (name: String, date: String, icon: String, color: Int?) -> Unit,
) {
    var name by remember { mutableStateOf(initialName) }
    var date by remember { mutableStateOf(initialDate) }
    var icon by remember { mutableStateOf(initialIcon) }
    var color by remember { mutableStateOf(initialColor ?: AppColorPalette.first()) }
    val overrideColor = hexColor(color)

    // iOS-parity birthday sheet (BirthdaySheet): name, date, icon grid, colour row.
    CreationSheet(
        title = title,
        confirmTitle = confirmLabel,
        confirmEnabled = name.isNotBlank() && date.isNotBlank(),
        onDismiss = onDismiss,
        onConfirm = { onConfirm(name.trim(), date, icon, color) },
    ) {
        Column(
            Modifier.verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            SheetField(
                icon = Icons.Filled.Person,
                placeholder = stringResource(R.string.name),
                value = name,
                onValueChange = { name = it },
            )
            BirthdayPickerField(value = date, onChange = { date = it }, label = stringResource(R.string.birthday))

            SectionLabel(stringResource(R.string.icon))
            IconGrid(
                options = IconOptions.calendar,
                selected = icon,
                onSelect = { icon = it },
                feature = FeatureAccent.Birthdays,
                colorOverride = overrideColor,
            )

            SectionLabel(stringResource(R.string.color))
            ColorPickerRow(selected = color, onSelect = { color = it })
        }
    }
}

@Composable
private fun SectionLabel(text: String) {
    Text(
        text.uppercase(),
        style = MaterialTheme.typography.labelMedium,
        fontWeight = FontWeight.SemiBold,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}
